openproblems/tasks/spatial_decomposition/datasets/_utils.py

```diff
@@ -0,0 +1,88 @@
+from .._utils import obs_means
+from anndata import AnnData
+
+import numpy as np
+import random
+
+
+def generate_seq(length=14):
+    return "".join(random.choice("CGTA") for _ in range(length))
+
+
+# pass the reference data
+def generate_synthatic_data(adata, sim_type="avg"):
+    n_genes = adata.n_var
+    n_cells = adata.n_obs
+    n_types = len(set(adata.obs["label"].values))
+
+    # TODO(make these arguments)
+    bead_depth = 1000
+    num_of_beads = n_cells * 2
+    # generate proportion values
+    props = np.random.dirichlet(np.ones(n_types), num_of_beads)
+
+    true_proportion = np.zeros((num_of_beads, n_types))
+    bead_to_gene_matrix = np.zeros((num_of_beads, n_genes))
+
+    # if sim_type avg
+    # generate from avg profiles
+    if sim_type == "avg":
+        profile_mean = obs_means(adata, "label")
+        # run for each bead
+        for bead_index in range(num_of_beads):
+            allocation = np.random.multinomial(
+                bead_depth, props[bead_index, :], size=1
+            )[0]
+            true_proportion[bead_index, :] = allocation.copy()
+            for j in range(n_types):
+                gene_exp = np.random.multinomial(
+                    allocation[j], profile_mean.X[j, :], size=1
+                )[0]
+
+                bead_to_gene_matrix[bead_index, :] += gene_exp.copy()
```

Comment:
do we really need a copy here? I don't think you have to worry about about referencing the same object later on since this is arithmetics.

Reply:
that's true, for summing we don't.